geth/jail/jail.go

```diff
@@ -4,251 +4,266 @@ import (
+		return newJailErrorResponse(err)
 	}
 
-	res, err := cell.Call("call", nil, this, args)
+	value, err := cell.Get("catalog")
+	if err != nil {
+		return newJailErrorResponse(err)
+	}
 
-	return makeResult(res.String(), err)
+	return newJailResultResponse(value)
 }
 
-// Send is a wrapper for executing RPC calls from within Otto VM.
-// It uses own jail's VM instance instead of cell's one to
-// increase safety of cell's vm usage.
-// TODO(divan): investigate if it's possible to do conversions
-// withouth involving otto code at all.
-// nolint: errcheck, unparam
-func (jail *Jail) Send(call otto.FunctionCall) otto.Value {
-	request, err := jail.vm.Call("JSON.stringify", nil, call.Argument(0))
-	if err != nil {
-		throwJSException(err)
-	}
+func (j *Jail) cell(chatID string) (*Cell, error) {
+	j.cellsMx.RLock()
+	defer j.cellsMx.RUnlock()
 
-	rpc := jail.nodeManager.RPCClient()
-	// TODO(divan): remove this check as soon as jail cells have
-	// proper cancellation mechanism implemented.
-	if rpc == nil {
-		throwJSException(fmt.Errorf("Error getting RPC client. Node stopped?"))
+	cell, ok := j.cells[chatID]
+	if !ok {
+		return nil, fmt.Errorf("cell '%s' not found", chatID)
 	}
-	response := rpc.CallRaw(request.String())
 
-	// unmarshal response to pass to otto
-	var resp interface{}
-	err = json.Unmarshal([]byte(response), &resp)
+	return cell, nil
+}
+
+// Cell returns a cell by chatID. If it does not exist, error is returned.
+// Required by the Backend.
+func (j *Jail) Cell(chatID string) (common.JailCell, error) {
+	return j.cell(chatID)
+}
+
+// Call executes the `call` function within a cell with chatID.
+// Returns a string being a valid JS code. In case of a successful result,
+// it's {"result": any}. In case of an error: {"error": "some error"}.
+//
+// Call calls commands from `_status_catalog`.
+// commandPath is an array of properties to retrieve a function.
+// For instance:
+//   `["prop1", "prop2"]` is translated to `_status_catalog["prop1"]["prop2"]`.
+func (j *Jail) Call(chatID, commandPath, args string) string {
+	cell, err := j.cell(chatID)
 	if err != nil {
-		throwJSException(fmt.Errorf("Error unmarshalling result: %s", err))
+		return newJailErrorResponse(err)
 	}
-	respValue, err := jail.vm.ToValue(resp)
+
+	value, err := cell.Call("call", nil, commandPath, args)
 	if err != nil {
-		throwJSException(fmt.Errorf("Error converting result to Otto's value: %s", err))
+		return newJailErrorResponse(err)
 	}
 
-	return respValue
+	return newJailResultResponse(value)
 }
 
-func newErrorResponse(msg string, id interface{}) map[string]interface{} {
-	// Bundle the error into a JSON RPC call response
-	return map[string]interface{}{
-		"jsonrpc": "2.0",
-		"id":      id,
-		"error": map[string]interface{}{
-			"code":    -32603, // Internal JSON-RPC Error, see http://www.jsonrpc.org/specification#error_object
-			"message": msg,
-		},
+// GetRPCClient returns an rpc.Client.
+func (j *Jail) GetRPCClient() *rpc.Client {
```

Comment:
`GetRPCClient` -> `RPCClient`? Generally, should it implement `RPCClientProvider`?